test/jasmine/tests/select_test.js

```diff
@@ -109,6 +117,582 @@ var BOXEVENTS = [1, 2, 1];
+        var mock = require('@mocks/histogram_colorscale.json');
+        var firstBinPts = [0];
+        var secondBinPts = [1, 2];
+        var thirdBinPts = [3, 4, 5];
+
+        mock.layout.clickmode = 'event+select';
+        Plotly.plot(gd, mock.data, mock.layout)
+          .then(function() {
+              return clickFirstBinImmediately();
+          })
+          .then(function() {
+              assertSelectedPoints(firstBinPts);
+              return shiftClickSecondBin();
+          })
+          .then(function() {
+              assertSelectedPoints([].concat(firstBinPts, secondBinPts));
+              return shiftClickThirdBin();
+          })
+          .then(function() {
+              assertSelectedPoints([].concat(firstBinPts, secondBinPts, thirdBinPts));
+              return clickFirstBin();
+          })
+          .then(function() {
+              assertSelectedPoints([].concat(firstBinPts));
+              clickFirstBin();
+              return deselectPromise;
+          })
+          .then(function() {
+              assertSelectionCleared();
+          })
+          .catch(failTest)
+          .then(done);
+
+        function clickFirstBinImmediately() { return _immediateClickPt({ x: 141, y: 358 }); }
+        function clickFirstBin() { return _click(141, 358); }
+        function shiftClickSecondBin() { return _click(239, 330, { shiftKey: true }); }
+        function shiftClickThirdBin() { return _click(351, 347, { shiftKey: true }); }
+    });
+
+    it('@flaky ignores clicks on boxes in a box trace type', function(done) {
+        var mock = Lib.extendDeep({}, require('@mocks/box_grouped_horz.json'));
+
+        mock.layout.clickmode = 'event+select';
+        mock.layout.width = 1100;
+        mock.layout.height = 450;
+
+        Plotly.plot(gd, mock.data, mock.layout)
+          .then(function() {
+              return clickPtImmediately();
+          })
+          .then(function() {
+              assertSelectedPoints(2);
+              clickPt();
+              return deselectPromise;
+          })
+          .then(function() {
+              assertSelectionCleared();
+              clickBox();
+          })
+          .then(function() {
+              // TODO Be sure this is called "late enough" after clicking on box has been processed
+              // Maybe plotly_click event would get fired after any selection events?
+              assertSelectionCleared();
+          })
+          .catch(failTest)
+          .then(done);
+
+        function clickPtImmediately() { return _immediateClickPt({ x: 610, y: 342 }); }
+        function clickPt() { return _clickPt({ x: 610, y: 342 }); }
+        function clickBox() { return _clickPt({ x: 565, y: 329 }); }
+    });
+
+    describe('is disabled when clickmode does not include \'select\'', function() {
+        // TODO How to test for pan and zoom mode as well? Note, that
```

Comment:
@etpinard or @alexcjohnson Would you mind to take a look at this question? Maybe it's just my lack of understanding how Plotly works, but I wonder how I can be sure when a click has been processed when the click itself should not lead to any change.

Reply:
Hmm. I'm not entirely sure I understand your question here. Under dragmode `pan` and `zoom` and clickmode `event`, the "selection code" shouldn't be executed.

>  I could try to handle the plotly_click event

So yeah, asserting that `plotly_click` returns some info about `mock14Pts[1]` would suffice.

Now, if `clickmode: 'select'` both `plotly_click` and `plotly_selected` aren't fired under dragmode `pan` and `zoom`. To check that indeed the selection _happened_, I would simply check

```js
expect(gd.data[0].selectedpoints).toEqual([14]);
```

I hope this helps!

Reply:
Let me try to ask the question in a different way: I want to test that click-to-select is disabled when clickmode is `event`. 

So the thing is, when a click event happens, plotly.js is dealing with that click. But maybe in an asynchronous fashion, for example because of throttling. The question is, when can I be sure plotly.js has finished the asynchronous processing of the click.

What I want to avoid is the following:
1. emulating the click
2. asserting selection state hasn't changed (too early!)
3. plotly.js asynchronous handling of the click event runs and potentially changes selection state because of a bug

Reply:
I fixed those tests and in addition added some more that ensure plotly_click is triggered after plotly_selected.

Reply:
Nicely done. Conversation resolved.